sdk-lib/src/androidTest/java/cash/z/ecc/android/sdk/db/DatabaseCoordinatorTest.kt

```diff
@@ -32,47 +33,113 @@ class DatabaseCoordinatorTest {
             )
+            // Check if the second job hasn't finished first, and thus makes this test invalid
+            if (testResult.resultFile != null) {
+                testResult.firstJobValidity = false
+                return@launch
+            }
+            testResult.resultFile = firstResultFile
         }
-        val job2 = launch {
-            delay(1001)
-            testResult = dbCoordinator.cacheDbFile(
+
+        // Small delay to support the need for running the second job right after the first one starts. Note this
+        // works well on local emulators, but not on CI emulators.
+        // delay(1)
+
+        val secondJob = launch(start = CoroutineStart.UNDISPATCHED) {
+            // We check here if the first job is still running and its result is not already proceeded
+            if (!firstJob.isActive || testResult.resultFile != null) {
+                testResult.secondJobValidity = false
+                return@launch
+            }
+            testResult.secondJobValidity = true
+            testResult.resultFile = dbCoordinator.cacheDbFile(
                 ZcashNetwork.Mainnet,
                 "TestZcashSdk"
             )
         }
 
-        advanceTimeBy(1002)
+        // Wait until both jobs are done
+        joinAll(firstJob, secondJob)
 
-        job2.join().also {
-            assertTrue(testResult != null)
-            assertTrue(testResult!!.absolutePath.isNotEmpty())
-            assertTrue(testResult!!.absolutePath.contains(ZcashNetwork.Mainnet.networkName))
-            assertTrue(testResult!!.absolutePath.contains("TestZcashSdk"))
+        // Check validity of this test iteration
+        if (!testResult.firstJobValidity ||
+            !testResult.secondJobValidity
+        ) {
+            return@runBlocking false
         }
+
+        // And here we assert that the second job was the one which waited, and thus accessed as second and changed
+        // the shared result correctly
+        assertTrue(testResult.resultFile != null)
+        assertTrue(testResult.resultFile!!.absolutePath.isNotEmpty())
+        assertTrue(testResult.resultFile!!.absolutePath.contains(ZcashNetwork.Mainnet.networkName))
+        assertTrue(testResult.resultFile!!.absolutePath.contains("TestZcashSdk"))
+
+        return@runBlocking true
     }
 
+    /**
+     * This stress test runs its subtest repeatedly and counts its failed iterations. If it exceeds the allowed fail
+     * threshold, then we claim this test as unsuccessful.
+     *
+     * According to manual testing, the valid tests ratio is around 90-99% when running on local emulators or
+     * physical devices. But on CI we get about 60-70% valid tests ratio.
+     */
     @FlakyTest
     @Test
     @MediumTest
     fun mutex_stress_test() {
-        // We run the mutex test multiple times sequentially to catch a possible problem.
-        for (x in 0..9) {
-            mutex_test()
+        val allAttempts = 100
```

Comment:
I have a concern that the test fails on CI at all.  If our code is behaving correctly, then this should never fail.  This isn't like a UI test that could be flakey, but rather seems like some kind of non-determinism we don't yet understand.

Do you know where specifically the failures are happening?

Reply:
Yes, I think I know what happens here. We need to start the second job right after the first was started, but not before it to be able to test the mutex implementation.

Because I don't know any way how to hook on the moment when the first job was just started, I try to run them in parallel and check inside them if these conditions were applied. So in the first one I check if the second one already didn't write the result, and in the second one I check if the first one is still running. If any of these conditions do not apply, then I claim that the test is not valid and run the next iteration of it. But it does not mean that it failed; it just not met our conditions to be able to check mutex implementation in the DatabaseCoordinator.